32-ljedd2

LJEDD2 opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
###  **백준 - 투 포인터**  | [배열 합치기](https://www.acmicpc.net/problem/11728)

**문제 설명** 
- 정렬되어있는 두 배열 A와 B가 주어진다. 
- 두 배열을 합친 다음 정렬해서 출력하기.

**제한 조건**
- 배열 A의 크기 N, 배열 B의 크기 M이 주어진다. (1 ≤ N, M ≤ 1,000,000)
- 시간 제한 1.5초 

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
1시간 

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->
<!-- 알고리즘에 대한 지식이 전혀 없는 사람이 봐도 이해할 수 있도록 작성해주세요. 시각자료를 이용하면 더 좋습니다. -->

### 1. 문제 지문 읽기 
- '정렬되어 있는 두 리스트의 합집합' 문제 
- 리스트의 모든 원소를 합쳐서 정렬한 결과를 계산하기 

#### 1-1) 단순 정렬 및 합치기로 해결 
- 정렬 함수를 사용하여 접근할 수 있다. 

```python
arr = sorted(a + b)
print(*arr, sep=' ')
```

### 2. 투 포인터로 해결 
>[!NOTE] 
**투 포인터 알고리즘이란?**
- 리스트에 순차적으로 접근해야 할 때 두 개의 점의 위치를 기록하면서 처리하는 알고리즘 
- 1차원 배열이 있고 이 배열에서 각자 다른 원소를 가리키고 있는 2개의 포인터를 조작해가면서 원하는 것을 얻는 형태이다. 
- 정렬되어있는 두 리스트의 합집합에도 사용된다. 
- 병합정렬(merge sort)의 counquer 영역의 기초가 되기도 한다.

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-4/assets/78216102/fb913e92-a552-464d-9427-f2df3c908da7)

**💡 투포인터로 문제를 해결하는 방법**
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-4/assets/78216102/21e3add7-9d10-49ed-a551-804b7360af80)

**1) 배열 A의 현재 인덱스를 가리키는 a_idx , 배열 B의 현재 인덱스를 가리키는 b_idx 를 포인터로 둔다.** 

**2) a_idx와 b_idx가 끝까지 이동할 때까지 무한루프를 돈다.** 
- 각 위치의 숫자를 비교하여 작은 값을 result에 넣고 해당 인덱스 번호를 `++1` 해준다. 
- 리스트의 끝까지 이동할 때까지 반복한다. 
```python
    if a[a_idx] < b[b_idx]:
        result.append(a[a_idx])
        a_idx += 1

    else:
        result.append(b[b_idx])
        b_idx += 1
```
**3) 한 쪽이 처리할 요소가 존재하지 않을 때** 
- 반복문을 돌다보면 a와 b의 길이가 다른 경우, 어느 한쪽 리스트 숫자가 모두 사용되었다면 인덱스 에러가 발생하게 된다. 
- 따라서 남아있는 다른 리스트의 숫자를 찾아 계속 담아야 한다.
    -   a_idx가 n이면 ? b의 b_idx 값을 계속해서 result에 넣어준다.
    -   b_idx가 n이면 ? a의 a_idx 값을 계속해서 result에 넣어준다.

```python
    # 한쪽이 이미 정렬이 끝난 경우 
    if a_idx == n:
        result.append(b[b_idx])
        b_idx += 1
        continue

    elif b_idx == m :
        result.append(a[a_idx])
        a_idx += 1
        continue
```


### 3. 전체 코드 
```python
n, m = map(int,input().split())
a = list(map(int,input().split()))
b = list(map(int,input().split()))

a_idx = 0
b_idx = 0
result = []

# 최대 길이를 n,m 둘 다 달성하면 탈출 
while a_idx != n or b_idx != m:

    # 한쪽이 이미 정렬이 끝난 경우 
    if a_idx == n:
        result.append(b[b_idx])
        b_idx += 1
        continue

    elif b_idx == m :
        result.append(a[a_idx])
        a_idx += 1
        continue

    # 투포인터 정렬 

    if a[a_idx] < b[b_idx]:
        result.append(a[a_idx])
        a_idx += 1

    else:
        result.append(b[b_idx])
        b_idx += 1

print(*result)
```
## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

LJEDD2 committed:
2024-03-25 돌 게임 2

LJEDD2 committed:
2024-03-25

LJEDD2 committed:
2024-03-28 파스칼 삼각형

LJEDD2 committed:
2024-03-28

LJEDD2 committed:
2024-04-01 도시와 비트코인

LJEDD2 committed:
2024-04-01

LJEDD2 committed:
2024-04-03 배열 합치기

LJEDD2 committed:
2024-04-03

9kyo-hwang commented on LJEDD2/투 포인터/배열 합치기.py:
```diff
@@ -0,0 +1,32 @@
+result = []
+
+# 최대 길이를 n,m 둘 다 달성하면 탈출 
+while a_idx != n or b_idx != m:
```
a 인덱스, b 인덱스가 각각 n, m이 "아닌 동안"이란 조건보단 "n, m보다 작을 때"란 조건이 조금 더 보기 좋을 것 같네요.
여기서는 별 의미없긴 한데, 다른 문제같은 곳에서 n, m을 초과하는 경우 반복문이 다시 시행되기 때문에 리스트 인덱스를 벗어날 가능성이 존재합니다.
```py
while a_idx < n or b_idx < m:
    # ...
```

LJEDD2 replied:
오 ... 그렇군요 
프로그램을 도는 과정에서 논리적 오류가 발생할 여지가 있을 것 같네요 지적해주셔서 감사합니다 !! 참고하여 코드를 써보도록 하겠습니다 🤔🤔🤔

9kyo-hwang commented on LJEDD2/투 포인터/배열 합치기.py:
```diff
@@ -0,0 +1,32 @@
+# 최대 길이를 n,m 둘 다 달성하면 탈출 
+while a_idx != n or b_idx != m:
+    # 한쪽이 이미 정렬이 끝난 경우 
+    if a_idx == n:
+        result.append(b[b_idx])
+        b_idx += 1
+        continue
+
+    elif b_idx == m :
+        result.append(a[a_idx])
+        a_idx += 1
+        continue
```
제 경우엔 이 부분을 반복문 밖으로 빼는 걸 선호합니다.
메인 반복문에서는 두 리스트 중 하나를 처리하는 것에 중점을 두고, 그 다음에 남은 리스트 원소를 추가하는 식으로 합니다.
```py
while a_idx < n and b_idx < m:
    if A[a_idx] < B[b_idx]:
        # A 원소 추가
    else:
        # B 원소 추가

while a_idx < n:
    # A 리스트 남은 원소 result에 추가

while b_idx < m:
    # B 리스트 남은 원소 result에 추가
```

9kyo-hwang reviewed (approved):
디코에 투포인터 추천 문제 올려놨으니 그것도 츄라이츄라이~

janghw0126 reviewed (approved):
저도 처음에는 단순히 두 리스트를 합친 다음 정렬하여 출력하는 문제라고 해서 `sort`를 이용해서 풀어보았습니다 !
```python
import sys
input=sys.stdin.readline

N,M=map(int,input().split())

A=list(map(int,input().split()))
B=list(map(int,input().split()))

AB_list=A+B
AB_list=sorted(AB_list)

for char in AB_list:
    print(char,end=" ")
```
정은님의 수도코드를 보고 투 포인터로 해결할 수 있다는 것도 알게 되었네요 😮
또또 합병정렬의 conquer 영역에 사용된다는 것도 알게되었어요 ! 저번 중간고사 공부를 하면서 합병정렬에서 두 리스트를 합병할 때의 원리가 이해가 잘 안되었는데 이번에 확실하게 알게 되었습니다 😍

정은님의 수도코드를 보고 쀨받아서 투 포인터를 재귀로 한번 풀어봤는데 배열 A와 B의 크기가 애초에 최대 1000000이라서 재귀 깊이 제한 문제로 런타임 에러가 애초에 발생하는 코드였더라구요,,혹시 몰라서 코드 공유합니댜.,혹시혹시 해결방법이 있으면 알려주세여,,!! 🥹🥹
```python
def merge_sort_array(a, b, a_idx=0, b_idx=0, result=None):
    if result is None:
        result = []

    # 모든 요소를 result에 다 추가하면 반환
    if a_idx == len(a) and b_idx == len(b):
        return result

    # a 배열이 모두 처리된 경우
    if a_idx == len(a):
        return result + b[b_idx:]

    # b 배열이 모두 처리된 경우
    if b_idx == len(b):
        return result + a[a_idx:]

    # 투포인터로 정렬
    if a[a_idx] < b[b_idx]:
        result.append(a[a_idx])
        return merge_sort_array(a, b, a_idx + 1, b_idx, result)
    else:
        result.append(b[b_idx])
        return merge_sort_array(a, b, a_idx, b_idx + 1, result)

# 입력 받기
n, m = map(int, input().split())
a = list(map(int, input().split()))
b = list(map(int, input().split()))

# 재귀 함수 호출
result = merge_sort_array(a, b)

# 결과 출력
print(*result)
```

LJEDD2 commented:
@janghw0126 
헉 재귀함수로 제한을 증가시키거나 가지치기를 생각해봤는데 아직 해결하지 못했습니다 ... 😭
재귀보다는 `heapq.merge`라는 파이썬 내장 함수를 활용하면 더 간단하게 해결할 수 있더라는 사실을 알게 되었습니다 ! 

```python
from heapq import merge

n, m = map(int, input().split())
a = list(map(int, input().split()))
b = list(map(int, input().split()))

result = merge(a, b)
print(*result)
```
이렇게 아주 간단히 문제를 해결할 수 있더라구요 

내부적으로 최소 힙(min heap)이 사용되어 큰 데이터셋에 대해서도 효율적으로 동작한다고 합니다 ! 
최소 힙이라는 개념을 한번 더 복습해봐야겠어요 :)
